### PR TITLE
bugfix dangling pointers

### PR DESCRIPTION
**Pull request by bertmelis**
adds ASYNC_WRITE_FLAG_COPY to copy data to LWIP avoiding dangling pointers.

small amounts of data are collected in a temporary buffer to be sent at once to (try to) avoid heap fragmentation.

### CODE REVIEW

**Comment by bertmelis** [comment]
Could anyone give this a review.

I do not see this as an endpoint but as a quick fix which could already solve a lot of problems. However, I do want to merge this and improve afterwards, together with better QoS management.

**Comment by kleini** [comment]
I don't understand why giving larger buffers to `ESPAsyncTCP.add()` should fix anything. I thought, sending smaller chunks of data would be better with devices having only very limited memory. Only advantage is, the `add()` method is not called that often.
I have this change running on some of my devices and it looks good, although I am using MQTT over SSL.

**Comment by bertmelis** [comment]
the PR is twofold:
- The COPY flag makes sure LWIP copies that data and only frees the memory when the TCP ack is received.
- Grouping all the `add` statements is only a (naive) way to prevent heap fragmentation.

Memory usage is higher than without the PR. When calling the eg `publish`, each message will be in memory twice: once in the temporary buffer and once in the LWIP sendqueue. The first one will be destroyed before exiting the mqtt-method. The other after receiving the TCP ACK by LWIP.

I did test this by sending 30 messages 100 bytes (QoS0) each on connect. Went fine.

**Comment by bertmelis** [comment]
PS I don't know the stack size (max is 4k?) so I took the safe path and create the temporary buffer on the heap to allow for larger payloads.

**Comment by kleini** [comment]
@bertmelis thanks for explaining the background of this PR. I had problems when homie-esp8266 with a lot of HomieNodes started up and it was not able to publish all messages necessary for all nodes over SSL. Will try to reproduce that and test than with this PR again.

**Comment by proddy** [comment]
I'll run some tests too. I'm already using a patched version from Bert's earlier changes and its working fine so don't expect any anomalies. The ESP8266 does have only 4KB stack which quickly gets used up especially if you're using the hogging Arduino libraries with arduinojson/aysncwebserver/asynctcp combo. My project has only 8KB heap left (from the 53K at boot) once all the libraries are loaded. If I had time I would port over to ESP-IDF or maybe go ESP32 only

**Comment by bertmelis** [comment]
> My project has only 8KB heap left (from the 53K at boot) once all the libraries are loaded.

8k? Hmmm, that's quite low. For example, the AsyncTCPBuffer as in the ESPAsyncTCP library doesn't allow allocation of the free heap is less then 4096. I was using the same limit to disallow new messages to be added to the outgoing queue I'm implementing for this lib.

File: src/AsyncMqttClient.cpp [hunk 1]
```diff
@@ -310,7 +310,11 @@ void AsyncMqttClient::_onConnect(AsyncClient* client) {
...
 
-  _client.add(fixedHeader, 1 + remainingLengthLength);
+  char* buff = new char[neededSpace];
+  size_t pos = 0;
+
+  memcpy(&buff[pos], fixedHeader, 1 + remainingLengthLength);
+  pos += 1 + remainingLengthLength;
 
   // Using a sendbuffer to fix bug setwill on SSL not working
   char sendbuffer[12];
```
Comment (mknjc) [inline src/AsyncMqttClient.cpp:320]: I would really like to see something like https://github.com/mknjc/async-mqtt-client/blob/04df4e00f55b0ffd7a6872c48c386428bdff924f/src/AsyncMqttClient/Headers.hpp#L7 to make it more readable

**Review by mknjc — CHANGES_REQUESTED** [review]
Instead adding a own layer of complexity what about using the lwip internal functionality.
lwip allocates the buffers to send (with the esp8266 config at least) in TCP_MSS size blocks and does the defragmentation by itself if told so with the apiflag `TCP_WRITE_FLAG_MORE`

**Comment by bertmelis** [comment]
> Instead adding a own layer of complexity what about using the lwip internal functionality.
> lwip allocates the buffers to send (with the esp8266 config at least) in TCP_MSS size blocks and does the defragmentation by itself if told so with the apiflag `TCP_WRITE_FLAG_MORE`

It this somewhere in the LWIP docs? I didn't (want to) check the source code.

**Comment by kleini** [comment]
> Instead adding a own layer of complexity what about using the lwip internal functionality.
> lwip allocates the buffers to send (with the esp8266 config at least) in TCP_MSS size blocks and does the defragmentation by itself if told so with the apiflag `TCP_WRITE_FLAG_MORE`

@mknjc You're invited to create pull requests. Let us all combine our efforts to make async-mqtt-client a better implementation. Your code looks really very readable. Let us try to get this into this library step by step.

File: src/AsyncMqttClient.cpp (hunk 1 above)
Comment (bertmelis) [reply src/AsyncMqttClient.cpp:320]: > I would really like to see something like https://github.com/mknjc/async-mqtt-client/blob/04df4e00f55b0ffd7a6872c48c386428bdff924f/src/AsyncMqttClient/Headers.hpp#L7 to make it more readable

Off course. I just wanted a quick bugfix.

**Comment by mknjc** [comment]
> It this somewhere in the LWIP docs? I didn't (want to) check the source code.

https://www.nongnu.org/lwip/2_1_x/group__lwip__opts__tcp.html#ga5648e2580bb55c0efdfbebcf3bad1eef
documents the general feature of lwip's pbuf chain reducing.
The arduino code has the feature enabled:
https://github.com/esp8266/Arduino/blob/a3a7a23a92508ce60c4ec0491dc3e0318978c8f5/tools/sdk/lwip2/include/lwipopts.h#L1418

The decision when the pbuf allocation should be oversized isn't documented afaik but is here in code https://git.savannah.nongnu.org/cgit/lwip.git/tree/src/core/tcp_out.c?h=STABLE-2_1_2_RELEASE#n243



> @mknjc You're invited to create pull requests. Let us all combine our efforts to make async-mqtt-client a better implementation. Your code looks really very readable. Let us try to get this into this library step by step.

I'm quite time constrained but if you all are interested I'm willing to invest time to open some merge requests.

**Comment by kleini** [comment]
> I'm quite time constrained but if you all are interested I'm willing to invest time to open some merge requests.

I think, we're all time constrained, right? I don't know, what @bertmelis thinks? But I would say, shared work is less work. And I see, that you put a lot of work into your fork. But your fork is not visible to "the rest" using async-mqtt-client from platformio.org registry. And I would like to get some async-mqtt-client, that works for larger payloads to be published, as I want to base my home automation on ESP8266 or ESP32 and MQTT and homie-esp8266. I want to create multi relay switches and push button devices, that integrate nicely into wall switches. I will publish PCBs for them.

**Comment by proddy** [comment]
I would say just create a `dev` branch in this repo and push all the new changes to it so we can all test and contribute. It doesn't have to work first time. The more we try the more stable it will become. And we all have different needs, some want support for larger payloads, some support for qos1/2 and me, I just want to fix my darn heap available space!

**Comment by bertmelis** [comment]
I don't have the powers to create a branch. I didn't try via CLI, but at least via the website I can't. Maybe @marvinroger can create a dev branch?

Meanwhile I created separate classes for all outgoing type of messages. So instead of creating a temp raw buffer, it is now wrapped in a class. The purpose of this is to have put them all in a queue so they don't have to be copied by LWIP. Reason for this is that LWIP discards the data once the ack is received while MQTT only can discard when the protocol acknowledges.

**Comment by kleini** [comment]
Count me in for testing the `dev` branch. I will try to find documentation, how LWIP, ESPAsyncTCP and async-mqtt-client work and try to understand the whole implementation.

**Comment by proddy** [comment]
@bertmelis @mknjc FYI I came across [AsyncTCPSock](https://github.com/yubox-node-org/AsyncTCPSock) which is an adaptation of the original AsyncTCP but re-written to use the built-in LWIP and uses the high-level BSD socket API. It's only for ESP32 at the moment but should give you an idea of how to use the LWIP library. His re-write focuses on removing the event queue that can fill up quickly with too many deadlocks and stops concurrency and replaces this with low-level select() calls to the network sockets & used to implement the callbacks.

**Comment by mknjc** [comment]
>  but re-written to use the built-in LWIP

At leased the AsyncTCP lib for arduino esp8266 uses the "built-in LWIP" provided by the arduino folks

>and uses the high-level BSD socket API.

the bsd socket API is really a compatibility layer above the netconn api which is build for blocking socket access, it uses a command queue and a tcp "thread". It's not really meant for small µcs with no task management and not so much memory for additional stacks.
The bsd socket api don't give any guarentees and for this reason the author also implemented it own in memory write cache which copies the data to send around again. (So there are at least two copies happening, one in lwip and one in the AsyncTCPSock lib)

>His re-write focuses on removing the event queue that

His rewrite added two queues into the loop, his own write buffer queue and the netconn command queue inside lwip

>replaces this with low-level select() calls to the network sockets & used to implement the callbacks.

select is one of the highest level functions in lwip and using it will cost already a lot of memory for its internal socket state machine

I'm quite sure switching to anything higher level then the currently used lwip api which is meant for low memory overhead async communication is a step into the wrong direction

**Comment by marvinroger** [comment]
> I don't have the powers to create a branch. I didn't try via CLI, but at least via the website I can't. Maybe @marvinroger can create a dev branch?
> 
> Meanwhile I created separate classes for all outgoing type of messages. So instead of creating a temp raw buffer, it is now wrapped in a class. The purpose of this is to have put them all in a queue so they don't have to be copied by LWIP. Reason for this is that LWIP discards the data once the ack is received while MQTT only can discard when the protocol acknowledges.

Of course, there you go! https://github.com/marvinroger/async-mqtt-client/tree/develop

**Comment by bertmelis** [comment]
I've been testing with a real sending queue:
- size of the data is only limited by memory
- data is saved into the queue per-message
- data can be sent in chunks
- upon disconnect, a complete message can be sent again (not just unacked data)
- messages are only discarded when the TCP ack arrives

I've made this with this lib in mind: https://github.com/bertmelis/AsyncTCPSendingQueue
It is still raw and unusable. But I've been thinking on how to best solve this lib's issues.

Any thoughts?

TO DO:
- make the classes extensible (to allow overriding methods)
- allow insertion at head of sending queue (to allow CONNECT message for MQTT)
- make it a proper lib
- multithreading protection has fatal bugs

**Comment by kleini** [comment]
> Any thoughts?

How can I help?

**Comment by bertmelis** [comment]
> > Any thoughts?
> 
> How can I help?

I'm still struggling with architecture. (I'm just an amateur you know) I'll post a first solution for review soon (later this week).

In any case: I'm building a std::list with std::vectors to hold data. And send that one by one.

**Comment by proddy** [comment]
oh be careful with std::list. each node takes a lot of heap memory overhead. Best use a std::deque if you want to add at both ends (front & back) and a char[] buffer since its a know set of bytes. But this is something we can help optimize when its all working.

**Comment by bertmelis** [comment]
I know, but in my design I need iterators to stay valid when inserting/removing. How to keep track what is sent or acked and what not? Sure I can just use a counter and do iterator maths on every operation...

Anyway, not optimizing yet, first make it work. you can follow here: https://github.com/bertmelis/TCPSendingQueue

**Comment by bertmelis** [comment]
Ok, I realized I'm overthinking. The ESP only has a very limited window size so I can (for now?) limit to send messages one after the other, waiting the send the next after the previous has been acked

That way, it's easier to guarantee message ordering, QoS>0 retries etc.

So yes, a simple deque will do. (Specifically say deque as it must be possible to insert at the front too, for the connection message.)

**Pull request closed without merging**